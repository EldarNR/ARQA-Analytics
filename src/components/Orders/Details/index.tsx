// type
import { OrderDetailsDialogProps } from '@/components/Orders/Details/type';
// components
import { Badge } from '@/components/ui/badge';
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
} from '@/components/ui/dialog';
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from '@/components/ui/table';
// helpers
import { getStatusBadgeVariant } from '@/helpers/getStatusBadgeVariant';
import { format } from 'date-fns';

export const OrderDetailsDialog = ({ order, customer, onClose }: OrderDetailsDialogProps) => {
  return (
    <Dialog onOpenChange={() => onClose()} open={!!order}>
      <DialogContent className="max-w-2xl">
        <DialogHeader>
          <DialogTitle>Order Details - {order?.id}</DialogTitle>
          <DialogDescription>Complete order information and items</DialogDescription>
        </DialogHeader>

        {order && (
          <div className="space-y-6">
            <div className="grid grid-cols-2 gap-4">
              <div>
                <h4 className="font-semibold mb-2">Order Information</h4>
                <div className="space-y-1 text-sm">
                  <div>
                    <strong>Date:</strong> {format(new Date(order.date), 'PPP')}
                  </div>
                  <div>
                    <strong>Customer:</strong> {customer?.name || order.customerId}
                  </div>
                  <div>
                    <strong>City:</strong> {order.city}
                  </div>
                  <div>
                    <strong>Channel:</strong> {order.channel}
                  </div>
                  <div>
                    <strong>Status:</strong>
                    <Badge className="ml-2" variant={getStatusBadgeVariant(order.status)}>
                      {order.status}
                    </Badge>
                  </div>
                </div>
              </div>

              <div>
                <h4 className="font-semibold mb-2">Total</h4>
                <div className="text-2xl font-bold text-green-600">
                  {new Intl.NumberFormat('kk-KZ', {
                    style: 'currency',
                    currency: 'KZT',
                    minimumFractionDigits: 0,
                  }).format(order.total)}
                </div>
                {order.comment && (
                  <div className="mt-4">
                    <h4 className="font-semibold mb-1">Comment</h4>
                    <p className="text-sm text-gray-600 bg-gray-50 p-2 rounded">{order.comment}</p>
                  </div>
                )}
              </div>
            </div>

            <div>
              <h4 className="font-semibold mb-3">Order Items</h4>
              <Table>
                <TableHeader>
                  <TableRow>
                    <TableHead>SKU</TableHead>
                    <TableHead>Product</TableHead>
                    <TableHead className="text-right">Qty</TableHead>
                    <TableHead className="text-right">Price</TableHead>
                    <TableHead className="text-right">Total</TableHead>
                  </TableRow>
                </TableHeader>
                <TableBody>
                  {order.items.map((item, index) => (
                    <TableRow key={index}>
                      <TableCell className="font-medium">{item.sku}</TableCell>
                      <TableCell>{item.name}</TableCell>
                      <TableCell className="text-right">{item.qty}</TableCell>
                      <TableCell className="text-right">
                        {new Intl.NumberFormat('kk-KZ', {
                          style: 'currency',
                          currency: 'KZT',
                          minimumFractionDigits: 0,
                        }).format(item.price)}
                      </TableCell>
                      <TableCell className="text-right font-medium">
                        {new Intl.NumberFormat('kk-KZ', {
                          style: 'currency',
                          currency: 'KZT',
                          minimumFractionDigits: 0,
                        }).format(item.price * item.qty)}
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            </div>
          </div>
        )}
      </DialogContent>
    </Dialog>
  );
};
