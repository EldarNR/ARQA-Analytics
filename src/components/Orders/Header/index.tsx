// type
import { OrdersTableHeaderProps } from '@/components/Orders/Header/type';
// components
import { TableHead, TableHeader, TableRow } from '@/components/ui/table';
// type
import { SortDirection, SortField } from '@/providers/type';
// icons
import { ChevronDownIcon, ChevronUpIcon } from 'lucide-react';

const SortIcon = ({
  field,
  sortField,
  sortDirection,
}: {
  field: SortField;
  sortField: SortField;
  sortDirection: SortDirection;
}) => {
  if (sortField !== field) {
    return null;
  }

  return sortDirection === 'asc' ? (
    <ChevronUpIcon className="h-4 w-4 text-gray-600 dark:text-gray-400" />
  ) : (
    <ChevronDownIcon className="h-4 w-4 text-gray-600 dark:text-gray-400" />
  );
};

export const OrdersTableHeader = ({ sortField, sortDirection, onSort }: OrdersTableHeaderProps) => {
  return (
    <TableHeader>
      <TableRow className="bg-gray-50 dark:bg-gray-800">
        <TableHead
          className="cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors
          text-gray-900 dark:text-gray-100 font-semibold"
          onClick={() => onSort('id')}
        >
          <div className="flex items-center gap-1">
            Order ID
            <SortIcon field="id" sortDirection={sortDirection} sortField={sortField} />
          </div>
        </TableHead>

        <TableHead
          className="cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors
          text-gray-900 dark:text-gray-100 font-semibold"
          onClick={() => onSort('date')}
        >
          <div className="flex items-center gap-1">
            Date
            <SortIcon field="date" sortDirection={sortDirection} sortField={sortField} />
          </div>
        </TableHead>

        <TableHead
          className="cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors
          text-gray-900 dark:text-gray-100 font-semibold"
          onClick={() => onSort('customerId')}
        >
          <div className="flex items-center gap-1">
            Customer
            <SortIcon field="customerId" sortDirection={sortDirection} sortField={sortField} />
          </div>
        </TableHead>

        <TableHead
          className="cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors
          text-gray-900 dark:text-gray-100 font-semibold"
          onClick={() => onSort('city')}
        >
          <div className="flex items-center gap-1">
            City
            <SortIcon field="city" sortDirection={sortDirection} sortField={sortField} />
          </div>
        </TableHead>

        <TableHead
          className="cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors
          text-gray-900 dark:text-gray-100 font-semibold"
          onClick={() => onSort('channel')}
        >
          <div className="flex items-center gap-1">
            Channel
            <SortIcon field="channel" sortDirection={sortDirection} sortField={sortField} />
          </div>
        </TableHead>

        <TableHead
          className="cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors
          text-gray-900 dark:text-gray-100 font-semibold"
          onClick={() => onSort('status')}
        >
          <div className="flex items-center gap-1">
            Status
            <SortIcon field="status" sortDirection={sortDirection} sortField={sortField} />
          </div>
        </TableHead>

        <TableHead
          className="cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700 transition-colors
          text-gray-900 dark:text-gray-100 font-semibold text-right"
          onClick={() => onSort('total')}
        >
          <div className="flex items-center justify-end gap-1">
            Total
            <SortIcon field="total" sortDirection={sortDirection} sortField={sortField} />
          </div>
        </TableHead>

        <TableHead className="w-[50px] text-gray-900 dark:text-gray-100 font-semibold">
          Actions
        </TableHead>
      </TableRow>
    </TableHeader>
  );
};
