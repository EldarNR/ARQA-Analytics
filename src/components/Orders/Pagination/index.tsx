// type
import { OrdersPaginationProps } from '@/components/Orders/Pagination/type';
// components
import { Button } from '@/components/ui/button';

export const OrdersPagination = ({
  currentPage,
  totalPages,
  itemsPerPage,
  totalItems,
  onPageChange,
}: OrdersPaginationProps) => {
  if (totalPages <= 1) {
    return null;
  }

  const startItem = (currentPage - 1) * itemsPerPage + 1;
  const endItem = Math.min(currentPage * itemsPerPage, totalItems);

  return (
    <div className="flex items-center justify-between mt-4">
      <div className="text-sm text-gray-600">
        Showing {startItem} to {endItem} of {totalItems} orders
      </div>
      <div className="flex gap-2">
        <Button
          disabled={currentPage === 1}
          onClick={() => onPageChange(Math.max(1, currentPage - 1))}
          variant="outline"
        >
          Previous
        </Button>
        <div className="flex gap-1">
          {[...Array(totalPages)].map((_, i) => (
            <Button
              key={i + 1}
              onClick={() => onPageChange(i + 1)}
              size="sm"
              variant={currentPage === i + 1 ? 'default' : 'outline'}
            >
              {i + 1}
            </Button>
          ))}
        </div>
        <Button
          disabled={currentPage === totalPages}
          onClick={() => onPageChange(Math.min(totalPages, currentPage + 1))}
          variant="outline"
        >
          Next
        </Button>
      </div>
    </div>
  );
};
