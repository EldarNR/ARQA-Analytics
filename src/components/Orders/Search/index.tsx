// type
import type { OrdersSearchProps } from '@/components/Orders/Search/type';
// components
import { Input } from '@/components/ui/input';
// icons
import { SearchIcon } from 'lucide-react';

export const OrdersSearch = ({ searchTerm, onSearchChange }: OrdersSearchProps) => {
  return (
    <div className="relative">
      <SearchIcon className="absolute left-3 top-3 h-4 w-4 text-gray-400" />
      <Input
        className="pl-10 w-64"
        onChange={(e) => onSearchChange(e.target.value)}
        placeholder="Search orders..."
        value={searchTerm}
      />
    </div>
  );
};
