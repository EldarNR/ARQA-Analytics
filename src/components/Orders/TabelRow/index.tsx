// type
import type { OrdersTableRowProps } from '@/components/Orders/TabelRow/type';
// components
import { Badge } from '@/components/ui/badge';
import { Button } from '@/components/ui/button';
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';
import { TableCell, TableRow } from '@/components/ui/table';
// helpers
import { getStatusBadgeVariant } from '@/helpers/getStatusBadgeVariant';
import { format } from 'date-fns';
// icons
import { EyeIcon, Loader2, MoreHorizontalIcon } from 'lucide-react';

export const OrdersTableRow = ({
  order,
  customer,
  onOrderClick,
  onStatusChange,
  isUpdating = false,
}: OrdersTableRowProps) => {
  return (
    <TableRow
      className="cursor-pointer hover:bg-gray-50 dark:hover:bg-gray-800 transition-colors"
      onClick={() => onOrderClick(order)}
    >
      <TableCell className="font-medium text-gray-900 dark:text-gray-100">{order.id}</TableCell>

      <TableCell className="text-gray-700 dark:text-gray-300">
        {format(new Date(order.date), 'MMM dd, yyyy')}
      </TableCell>

      <TableCell className="text-gray-700 dark:text-gray-300">
        {customer?.name || order.customerId}
      </TableCell>

      <TableCell className="text-gray-700 dark:text-gray-300">{order.city}</TableCell>

      <TableCell className="text-gray-700 dark:text-gray-300">{order.channel}</TableCell>

      <TableCell>
        <Badge variant={getStatusBadgeVariant(order.status)}>{order.status}</Badge>
      </TableCell>

      <TableCell className="text-right font-medium text-gray-900 dark:text-gray-100">
        {new Intl.NumberFormat('kk-KZ', {
          style: 'currency',
          currency: 'KZT',
          minimumFractionDigits: 0,
        }).format(order.total)}
      </TableCell>

      <TableCell onClick={(e) => e.stopPropagation()}>
        <DropdownMenu>
          <DropdownMenuTrigger asChild>
            <Button
              className="h-8 w-8 p-0 hover:bg-gray-100 dark:hover:bg-gray-700"
              disabled={isUpdating}
              variant="ghost"
            >
              {isUpdating ? (
                <Loader2 className="h-4 w-4 animate-spin" />
              ) : (
                <MoreHorizontalIcon className="h-4 w-4 text-gray-600 dark:text-gray-400" />
              )}
            </Button>
          </DropdownMenuTrigger>
          <DropdownMenuContent
            align="end"
            className="bg-white dark:bg-gray-800 border border-gray-200 dark:border-gray-700"
          >
            <DropdownMenuItem
              className="text-gray-900 dark:text-gray-100 hover:bg-gray-100 dark:hover:bg-gray-700 cursor-pointer"
              onClick={() => onOrderClick(order)}
            >
              <EyeIcon className="mr-2 h-4 w-4" />
              View Details
            </DropdownMenuItem>

            <DropdownMenuItem
              className="text-gray-900 dark:text-gray-100 hover:bg-gray-100 dark:hover:bg-gray-700 cursor-pointer disabled:opacity-50 disabled:cursor-not-allowed"
              disabled={order.status === 'New' || isUpdating}
              onClick={() => onStatusChange(order.id, 'New')}
            >
              Mark as New
            </DropdownMenuItem>

            <DropdownMenuItem
              className="text-gray-900 dark:text-gray-100 hover:bg-gray-100 dark:hover:bg-gray-700 cursor-pointer disabled:opacity-50 disabled:cursor-not-allowed"
              disabled={order.status === 'Processing' || isUpdating}
              onClick={() => onStatusChange(order.id, 'Processing')}
            >
              Mark as Processing
            </DropdownMenuItem>

            <DropdownMenuItem
              className="text-gray-900 dark:text-gray-100 hover:bg-gray-100 dark:hover:bg-gray-700 cursor-pointer disabled:opacity-50 disabled:cursor-not-allowed"
              disabled={order.status === 'Shipped' || isUpdating}
              onClick={() => onStatusChange(order.id, 'Shipped')}
            >
              Mark as Shipped
            </DropdownMenuItem>
          </DropdownMenuContent>
        </DropdownMenu>
      </TableCell>
    </TableRow>
  );
};
