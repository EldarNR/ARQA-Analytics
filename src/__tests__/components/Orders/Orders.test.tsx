import React from 'react';
import { render, screen } from '@testing-library/react';

import Orders from '@/app/orders/page';

// Моки для хуков
jest.mock('@/hooks/useOrders', () => ({
  useOrders: jest.fn(() => ({
    data: [],
    isLoading: false,
    error: null,
  })),
}));

jest.mock('@/hooks/useCustomers', () => ({
  useCustomers: jest.fn(() => ({
    data: [],
    isLoading: false,
  })),
}));

jest.mock('@/hooks/useUpdateOrderStatus', () => ({
  useUpdateOrderStatus: jest.fn(() => ({
    mutate: jest.fn(),
    isPending: false,
  })),
}));

jest.mock('@/components/Orders/Search', () => ({
  OrdersSearch: () => <div data-testid="search">Search</div>,
}));

// Моки для остальных компонентов
jest.mock('@/components/Orders/Details', () => ({
  OrderDetailsDialog: () => null,
}));

jest.mock('@/components/Orders/Header', () => ({
  OrdersTableHeader: () => (
    <thead data-testid="table-header">
      <tr>
        <th>Header</th>
      </tr>
    </thead>
  ),
}));

jest.mock('@/components/Orders/TabelRow', () => ({
  OrdersTableRow: () => (
    <tr data-testid="order-row">
      <td>Order</td>
    </tr>
  ),
}));

jest.mock('@/components/Orders/Pagination', () => ({
  OrdersPagination: () => <div data-testid="pagination">Pagination</div>,
}));

// Mock UI компонентов
jest.mock('@/components/ui/card', () => ({
  Card: ({ children }: any) => <div data-testid="card">{children}</div>,
  CardContent: ({ children }: any) => <div data-testid="card-content">{children}</div>,
  CardHeader: ({ children }: any) => <div data-testid="card-header">{children}</div>,
  CardTitle: ({ children }: any) => <h3 data-testid="card-title">{children}</h3>,
}));

jest.mock('@/components/ui/table', () => ({
  Table: ({ children }: any) => <table data-testid="table">{children}</table>,
  TableBody: ({ children }: any) => <tbody data-testid="table-body">{children}</tbody>,
}));

describe('Orders Component', () => {
  const mockUseOrders = require('@/hooks/useOrders').useOrders;
  const mockUseCustomers = require('@/hooks/useCustomers').useCustomers;

  beforeEach(() => {
    mockUseOrders.mockReturnValue({
      data: [],
      isLoading: false,
      error: null,
    });

    mockUseCustomers.mockReturnValue({
      data: [],
      isLoading: false,
    });
  });

  test('renders orders page', () => {
    render(<Orders />);
    expect(screen.getByText('Orders')).toBeInTheDocument();
  });

  test('shows search component', () => {
    render(<Orders />);
    expect(screen.getByTestId('search')).toBeInTheDocument();
  });

  test('shows loading state', () => {
    mockUseOrders.mockReturnValue({
      data: [],
      isLoading: true,
      error: null,
    });

    render(<Orders />);
    expect(screen.getByText('Orders')).toBeInTheDocument();
  });
});
