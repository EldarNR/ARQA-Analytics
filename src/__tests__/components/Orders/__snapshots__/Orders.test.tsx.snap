// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Orders Component Rendering matches snapshot with orders 1`] = `
<div
  class="space-y-4"
>
  <div
    class="flex justify-between items-center"
  >
    <h1
      class="text-2xl font-bold"
    >
      Orders
    </h1>
    <div
      class="flex items-center gap-4"
    >
      <input
        data-testid="search-input"
        placeholder="Search orders..."
        value=""
      />
    </div>
  </div>
  <div
    data-testid="card"
  >
    <div
      data-testid="card-header"
    >
      <h3
        data-testid="card-title"
      >
        Orders List (
        2
         orders)
      </h3>
    </div>
    <div
      data-testid="card-content"
    >
      <table
        data-testid="table"
      >
        <thead
          data-testid="table-header"
        >
          <tr>
            <th
              data-testid="sort-date"
            >
              Date 
              desc
            </th>
            <th
              data-testid="sort-customer"
            >
              Customer 
            </th>
            <th
              data-testid="sort-total"
            >
              Total 
            </th>
          </tr>
        </thead>
        <tbody
          data-testid="table-body"
        >
          <tr
            data-testid="order-row"
          >
            <td>
              order-2
            </td>
            <td>
              Jane Smith
            </td>
            <td>
              8500
            </td>
            <td>
              pending
            </td>
          </tr>
          <tr
            data-testid="order-row"
          >
            <td>
              order-1
            </td>
            <td>
              John Doe
            </td>
            <td>
              15000
            </td>
            <td>
              completed
            </td>
          </tr>
        </tbody>
      </table>
      <div
        data-testid="pagination"
      >
        Page 
        1
         of 
        1
        <button>
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
